src/groupeddataframe/grouping.jl

```diff
@@ -181,6 +197,319 @@ end
+_grouptypes(gd::GroupedDataFrame) = Tuple(eltype(gd.parent[!, c]) for c in gd.cols)
+
+
+"""
+    GroupKey
+
+Key for one of the groups of a [`GroupedDataFrame`](@ref). Contains the values
+of the corresponding grouping columns and behaves similarly to a `NamedTuple`,
+but using it to index its `GroupedDataFrame` is much more effecient than using the
+equivalent `Tuple` or `NamedTuple`.
+
+Instances of this type are returned by `keys(::GroupedDataFrame)` and are not
+meant to be constructed directly.
+
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKey{P<:GroupedDataFrame}
+    parent::P
+    idx::Int
+end
+
+function Base.show(io::IO, k::GroupKey)
+    print(io, "GroupKey")
+    show(io, NamedTuple(k))
+end
+
+Base.parent(key::GroupKey) = getfield(key, :parent)
+Base.length(key::GroupKey) = length(parent(key).cols)
+Base.names(key::GroupKey) = Tuple(groupvars(parent(key)))
+Base.keys(key::GroupKey) = names(key)
+Base.propertynames(::GroupKey) = names(key)
+Base.values(key::GroupKey) = groupvalues(Tuple, parent(key), getfield(key, :idx))
+
+Base.iterate(key::GroupKey) = iterate(key, 1)
+Base.iterate(key::GroupKey, i::Integer) = i <= length(key) ? (key[i], i + 1) : nothing
+
+Base.getindex(key::GroupKey, i::Integer) = groupvalues(parent(key), getfield(key, :idx), i)
+
+function Base.getindex(key::GroupKey, n::Symbol)
+    try
+        return groupvalues(parent(key), getfield(key, :idx), n)
+    catch e
+        throw(KeyError(n))
+    end
+end
+
+function Base.getproperty(key::GroupKey, p::Symbol)
+    try
+        return key[p]
+    catch e
+        error("$(typeof(key)) has no property $p")
+    end
+end
+
+# Base.:(==)(k1::GroupKey, k2::GroupKey) = parent(k1) === parent(k2) && getfield(k1, :idx) == getfield(k2, :idx)
+
+Base.NamedTuple(key::GroupKey) = groupvalues(NamedTuple, parent(key), getfield(key, :idx))
+Base.Tuple(key::GroupKey) = values(key)
+
+
+"""
+    keys(gd::GroupedDataFrame)
+
+Get the set of keys for each group in `GroupedDataFrame` object `gd`.
+The keys are [`GroupKey`](@ref) objects which behave like `NamedTuple`s holding
+the values of the grouping keys and can be used to index into `gd` efficiently.
+
+### Returns
+
+Iterator over [`GroupKey`](@ref)s.
+
+### Examples
+
+```jldoctest
+julia> df = DataFrame(a = repeat([:foo, :bar, :baz], outer=[4]),
+                      b = repeat([2, 1], outer=[6]),
+                      c = 1:12);
+
+julia> gd = groupby(df, [:a, :b])
+GroupedDataFrame with 6 groups based on keys: a, b
+First Group (2 rows): a = :foo, b = 2
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ foo    │ 2     │ 1     │
+│ 2   │ foo    │ 2     │ 7     │
+⋮
+Last Group (2 rows): a = :baz, b = 1
+│ Row │ a      │ b     │ c     │
+│     │ Symbol │ Int64 │ Int64 │
+├─────┼────────┼───────┼───────┤
+│ 1   │ baz    │ 1     │ 6     │
+│ 2   │ baz    │ 1     │ 12    │
+
+julia> collect(keys(gd))
```

Comment:
This doesn't sound like a good suggestion, better collect named tuples:
```suggestion
julia> NamedTuple.(keys(gd))
```

Also show the output of `keys(gd)` first.